tests/aws/app/app_test.go

```diff
@@ -0,0 +1,217 @@
+	"golang.org/x/crypto/ssh"
+	"google.golang.org/api/iterator"
+)
+
+const (
+	requestlogURL = "/requestlog/"
+	traceURL      = "/trace/"
+)
+
+var (
+	hostIP       string
+	awsRegion    string
+	gcpProjectID string
+	sshUser      string
+	sshKeyPath   string
+)
+
+func init() {
+	flag.StringVar(&hostIP, "host-ip", "", "host IP address of the instance")
+	flag.StringVar(&awsRegion, "aws-region", "us-west-1", "the region used to run the sample app")
+	flag.StringVar(&gcpProjectID, "gcp-project", "", "GCP project used to collect request logs")
+	flag.StringVar(&sshUser, "ssh-user", "admin", "user name used to ssh into the EC2 instance")
+	flag.StringVar(&sshKeyPath, "key-path", "", "path to the key file")
+}
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+	if hostIP == "" || gcpProjectID == "" {
+		log.Println("Test environments need to be setup to run server tests.")
+		flag.PrintDefaults()
+		return
+	}
+	if err := startApp(); err != nil {
+		log.Fatal("startApp:", err)
+	}
+	os.Exit(m.Run())
+}
+
+func startApp() error {
+	key, err := publicKeyFile(sshKeyPath)
+	if err != nil {
+		return fmt.Errorf("cannot get key from file: %v", err)
+	}
+	config := &ssh.ClientConfig{
+		User: sshUser,
+		Auth: []ssh.AuthMethod{key},
+		HostKeyCallback: func(hostname string, remote net.Addr, key ssh.PublicKey) error {
+			return nil
+		},
+		Timeout: 30 * time.Second,
+	}
+	c, err := ssh.Dial("tcp", hostIP+":22", config)
+	if err != nil {
+		return fmt.Errorf("failed to dial %s:22: %v", hostIP, err)
+	}
+	session, err := c.NewSession()
+	if err != nil {
+		return fmt.Errorf("failed to create session: %v", err)
+	}
+
+	// This close the session channel and the TCP connection, but doesn't terminate
+	// the running app. session.Signal() does not terminate it either, see
+	// https://github.com/golang/go/issues/4115. This is OK for our use case since
+	// we tear it down after the test finishes.
+	defer func() {
+		session.Close()
+		c.Close()
+	}()
+	return session.Start("/home/admin/app 2>&1 | logger")
+}
+
+func publicKeyFile(path string) (ssh.AuthMethod, error) {
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	key, err := ssh.ParsePrivateKey(b)
+	if err != nil {
+		return nil, err
+	}
+	return ssh.PublicKeys(key), nil
+}
+
+func TestRequestLog(t *testing.T) {
+	t.Parallel()
+	suf, err := testutil.URLSuffix()
+	if err != nil {
+		t.Fatal("cannot generate URL suffix:", err)
+	}
+	p := path.Clean(fmt.Sprintf("/%s/%s", requestlogURL, suf))
+	if err := testutil.Retry(t, testutil.Get("http://"+hostIP+":8080"+p)); err != nil {
+		t.Fatal(err)
+	}
+
+	ctx := context.Background()
+	c, err := logadmin.NewClient(ctx, gcpProjectID)
+	if err != nil {
+		t.Fatalf("error creating logadmin client: %v", err)
+	}
+	defer c.Close()
+
+	if err := testutil.Retry(t, readLogEntries(ctx, c, p)); err != nil {
+		t.Error(err)
+	}
+}
+
+func readLogEntries(ctx context.Context, c *logadmin.Client, u string) func() error {
```

Comment:
If you want to split this out into a separate function, don't return a `func() error`. Perform the code directly in the function, and change the caller to do the adapting into the right kind of closure.

Reply:
Done.